store/sqlstore/channel_store.go

```diff
@@ -2183,19 +2183,42 @@ func (s SqlChannelStore) SearchForUserInTeam(userId string, teamId string, term
 	})
 }
 
-func (s SqlChannelStore) SearchAllChannels(term string, opts store.ChannelSearchOpts) (*model.ChannelListWithTeamData, *model.AppError) {
+func (s SqlChannelStore) channelSearchQuery(term string, opts store.ChannelSearchOpts, countQuery bool) sq.SelectBuilder {
+	var limit int
+	if opts.PerPage != nil {
+		limit = *opts.PerPage
+	} else {
+		limit = 100
+	}
+
+	var selectStr string
+	if countQuery {
+		selectStr = "count(*)"
+	} else {
+		selectStr = "c.*, t.DisplayName AS TeamDisplayName, t.Name AS TeamName, t.UpdateAt as TeamUpdateAt"
+	}
+
 	query := s.getQueryBuilder().
-		Select("c.*, t.DisplayName AS TeamDisplayName, t.Name AS TeamName, t.UpdateAt as TeamUpdateAt").
+		Select(selectStr).
 		From("Channels AS c").
 		Join("Teams AS t ON t.Id = c.TeamId").
-		Where(sq.Eq{"c.Type": []string{model.CHANNEL_PRIVATE, model.CHANNEL_OPEN}}).
-		OrderBy("c.DisplayName, t.DisplayName").
-		Limit(uint64(100))
+		Where(sq.Eq{"c.Type": []string{model.CHANNEL_PRIVATE, model.CHANNEL_OPEN}})
+
+	// don't bother ordering or limiting if we're just getting the count
+	if !countQuery {
+		query = query.
+			OrderBy("c.DisplayName, t.DisplayName").
+			Limit(uint64(limit))
+	}
 
 	if !opts.IncludeDeleted {
 		query = query.Where(sq.Eq{"c.DeleteAt": int(0)})
 	}
 
+	if opts.Page != nil && opts.PerPage != nil && !countQuery {
```

Comment:
I'm actually going to add a method `(*ChannelSearchOpts).IsPaginated()` because this `nil` check logic is used a couple of times.